library/fluid/fluid_flow/src/main/java/org/quiltmc/qsl/fluid/flow/api/FluidFlowEvents.java

```diff
@@ -0,0 +1,116 @@
+ */
+
+package org.quiltmc.qsl.fluid.flow.api;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import org.jetbrains.annotations.Nullable;
+import org.quiltmc.qsl.base.api.event.Event;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.world.World;
+
+
+public final class FluidFlowEvents {
+	private FluidFlowEvents() {
+	}
+
+	// Flowing Block to Interacting Block to Direction to Event
+	private static final Map<Block, Map<Block, Map<Direction, Event<FluidFlowInteractionCallback>>>> EVENT_MAP = new HashMap<>();
+
+	/**
+	 * Registers a new event on a fluid flow. The same two blocks can register a different event for different directions, but the same directions will run the event that was registered first.
+	 *
+	 * @param flowingBlock          The fluid block that flowed.
+	 * @param interactionBlock      The block in one of the {@code interactionDirections}.
+	 * @param interactionDirections The direction to search for {@code interactionBlock}.
+	 * @param interactionEvent      The event to run when the conditions are met.
+	 */
+	public static void register(Block flowingBlock, Block interactionBlock, Direction[] interactionDirections, FluidFlowInteractionCallback interactionEvent) {
+		Map<Block, Map<Direction, Event<FluidFlowInteractionCallback>>> flowBlockEvents = EVENT_MAP.computeIfAbsent(flowingBlock, flowing -> new Object2ObjectOpenHashMap<>());
+		Map<Direction, Event<FluidFlowInteractionCallback>> interactionEvents = flowBlockEvents.computeIfAbsent(interactionBlock, interacting -> new Object2ObjectOpenHashMap<>());
+
+		// Create events for all the different directions for the specified blocks
+		if (interactionEvents.isEmpty()) {
+			for (Direction direction : Direction.values()) {
+				interactionEvents.put(direction, Event.create(FluidFlowInteractionCallback.class, fluidFlowInteractionEvents -> (flowingBlockState, interactingBlockState, flowPos, world) -> {
+					for (FluidFlowInteractionCallback event : fluidFlowInteractionEvents) {
+						if (!event.onFlow(flowingBlockState, interactingBlockState, flowPos, world)) {
+							return false;
+						}
+					}
+
+					return true;
+				}));
+			}
+		}
+
+		// Register the new event callbacks
+		for (Map.Entry<Direction, Event<FluidFlowInteractionCallback>> pair : interactionEvents.entrySet()) {
+			for (Direction direction : interactionDirections) {
+				if (pair.getKey() == direction) {
+					pair.getValue().register(interactionEvent);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Gets the event from the following blocks and direction.
+	 *
+	 * @param flowingBlock         The fluid block that flowed.
+	 * @param interactionBlock     The block it interacts with.
+	 * @param interactionDirection The interaction direction
+	 * @return An event if the conditions are met, otherwise {@code null}
```

Comment:
```suggestion
	 * @param flowingBlock         the fluid block that flowed
	 * @param interactionBlock     the block it interacts with
	 * @param interactionDirection the interaction direction
	 * @return an event if the conditions are met, otherwise {@code null}
```